### PR TITLE
Tick, Pools, and PoolReserves updates

### PR DESCRIPTION
**Pull request by NicholasDotSol**
Update documentation in advance of the DEX upgrade.

### CODE REVIEW

**Review by jcompagni10 — APPROVED** [review]

File: docs/neutron/modules/dex/overview/concepts/swaps.md [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
 
 1. All available liquidity has been exhausted.
-    - If there is no available tokenOut liquidity for the given pair at the beginning of the swap it will fail and return an `ErrInsufficientLiquidity`.
+    - If there is no available `tokenOut` liquidity for the given pair at the beginning of the swap it will fail and return an `ErrLimitPriceNotSatisfied`.
     - If swap is called through a `IMMEDIATE\_OR\_CANCEL` limit order it will still succeed if liquidity is exhausted and only a portion of the `AmountIn` has been used. In all other cases a partial fill of a swap will result in a failure.
 2. The `AmountIn` has been hit (ie. the user has swapped through 100% of the supplied `TokenIn`.
 3. `MaxAmountOut` has been set and the `TokenOut` amount is equal to `MaxAmountOut`.
     - In cases where only a portion of the maxAmountIn is used only the used portion of `TokenIn` will be debited from the user's account.
+x
```
Comment (pr0n00gler) [inline docs/neutron/modules/dex/overview/concepts/swaps.md:13]: I think this `x` should be removed

File: docs/neutron/modules/dex/overview/concepts/swaps.md [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
 
 1. All available liquidity has been exhausted.
-    - If there is no available tokenOut liquidity for the given pair at the beginning of the swap it will fail and return an `ErrInsufficientLiquidity`.
+    - If there is no available `tokenOut` liquidity for the given pair at the beginning of the swap it will fail and return an `ErrLimitPriceNotSatisfied`.
     - If swap is called through a `IMMEDIATE\_OR\_CANCEL` limit order it will still succeed if liquidity is exhausted and only a portion of the `AmountIn` has been used. In all other cases a partial fill of a swap will result in a failure.
 2. The `AmountIn` has been hit (ie. the user has swapped through 100% of the supplied `TokenIn`.
 3. `MaxAmountOut` has been set and the `TokenOut` amount is equal to `MaxAmountOut`.
     - In cases where only a portion of the maxAmountIn is used only the used portion of `TokenIn` will be debited from the user's account.
```
Comment (pr0n00gler) [inline docs/neutron/modules/dex/overview/concepts/swaps.md:12]: This line has a pretty weird style
Screenshot 2024-06-12 at 2 47 33 PM
, maybe change to something else?

**Review by pr0n00gler — CHANGES_REQUESTED** [review]

File: docs/neutron/modules/dex/overview/concepts/swaps.md (hunk 2 above)
Comment (jcompagni10) [reply docs/neutron/modules/dex/overview/concepts/swaps.md]: Not sure I understand what the issue is here. I did switch to clearer variables and add a missing coma. Let me know if there's something else I am missing @pr0n00gler .

**Review by jcompagni10 — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Review by NeverHappened — CHANGES_REQUESTED** [review]
lgtm, found a few typos and optional suggestions

File: docs/neutron/modules/dex/overview/concepts/amms-and-orderbooks.md [hunk 3]
```diff
@@ -7,23 +7,20 @@ AMMs were the first time that capital markets were created on-chain, in a comput
...
 
 This simple design has significant consequences. Neutron DEX can reach order-book levels of capital efficiency (zero price impact on trades, swaps, and limit orders) while still maintaining the computational efficiency and liveness properties of AMMs.
 
-
-
 **A Quick Note on Fees and Value Accrual**
 
 Concentrated liquidity and orderbooks differ slightly in how liquidity providers accrue value. On concentrated liquidity AMMs, liquidity providers accrue value through a [fee](docs/neutron/modules/dex/overview/concepts/fees.md) that they choose when placing the liquidity. The fee is a predetermined premium that traders pay when using the liquidity to swap between two assets.
 
 Orderbook liquidity providers accrue value through maintaining a bid-ask spread, which entails them offering to sell assets slightly above the current current price and buy them slightly below the current price. It turns out that these two mechanism are actually very similar in that they describe the difference in prices between the best sell and buy prices.
```
Comment (quasisamurai) [inline docs/neutron/modules/dex/overview/concepts/amms-and-orderbooks.md:26]:
```suggestion
Orderbook liquidity providers accrue value through maintaining a bid-ask spread, which entails them offering to sell assets slightly above the current price and buy them slightly below the current price. It turns out that these two mechanism are actually very similar in that they describe the difference in prices between the best sell and buy prices.
```

File: docs/neutron/modules/dex/overview/concepts/liquidity-iteration.md [hunk 4]
```diff
@@ -1,16 +1,66 @@
...
-- `PoolReserves`: In Ascending `Fee` order
-- `LimitOrderTranches`: In ascending `TrancheKey` order
+When swapping through liquidity via a Multi-Hop Swap or a Taker Limit Order we iterate through the available `TickLiquidity` to fill the swap order. Liquidity is always iterated through in order of best to worst price (from the taker's perspective.) For each swap, we completely exhaust the available reserves before moving on to the next tick. For `TickLiquidity` instances at the same `TickIndex` they are iterated through in a deterministic order as follows:
```
Comment (quasisamurai) [inline docs/neutron/modules/dex/overview/concepts/liquidity-iteration.md:3]:
```suggestion
When swapping through liquidity via a Multi-Hop Swap or a Taker Limit Order, we iterate through the available `TickLiquidity` to fill the swap order. Liquidity is always iterated through in order of best to worst price (from the taker's perspective.) For each swap, we completely exhaust the available reserves before moving on to the next tick. For `TickLiquidity` instances at the same `TickIndex` they are iterated through in a deterministic order as follows:
```

File: docs/neutron/modules/dex/overview/concepts/liquidity-iteration.md [hunk 5]
```diff
@@ -1,16 +1,66 @@
...
+
+1. The first available `TickLiquidity` holding `ATOM` is a `LimitOrderTranche` at tick $19640$. Since this is a limit order, when swapped through and depleted the liquidity is removed from state and the `USDC` Alice paid is credited to the limit order `Receiver`
+2. Alice swaps up to $71.428$ `USDC` using this pool before depleting the reserves. This will net her $10$ `ATOM` for the swap:
+
+ $\text{ATOM available} / \text{exchange rate} = \text{USDC needed}$
+ $10 / 0.14 = 71.428$
+
+
+3. Alice still has $28.527$ `USDC` she needs to swap, so we move to the next available tick: $-20795$
+This tick offers `ATOM` at a price of $0.123$ `ATOM` per `USDC` and is of type `PoolReserve`. Any USDC alice pays for this swap will be placed in the corresponding poolReserves of the Pool (USDC @ Tick $$-20795$$ .) Alice swaps the remainder of her `USDC` here, resulting in an additional $3.5$ `ATOM`:
```
Comment (quasisamurai) [inline docs/neutron/modules/dex/overview/concepts/liquidity-iteration.md:49]:
```suggestion
This tick offers `ATOM` at a price of $0.123$ `ATOM` per `USDC` and is of type `PoolReserve`. Any USDC Alice pays for this swap will be placed in the corresponding poolReserves of the Pool (USDC @ Tick $$-20795$$ .) Alice swaps the remainder of her `USDC` here, resulting in an additional $3.5$ `ATOM`:
```

**Review by quasisamurai — CHANGES_REQUESTED** [review]

File: docs/neutron/modules/dex/messages.md [hunk 6]
```diff
@@ -262,82 +327,43 @@ message MsgWithdrawFilledLimitOrder {
...
-    repeated uint64 fee_tiers = 1;
-    string max_true_taker_spread = 2 [
-        (gogoproto.moretags) = "yaml:\"max_true_taker_spread\"",
-        (gogoproto.customtype) = "github.com/neutron-org/neutron/v2/utils/math.PrecDec",
-        (gogoproto.nullable) = false,
-        (gogoproto.jsontag) = "max_true_taker_spread"
-    ];
+  option (gogoproto.goproto_stringer) = false;
+  repeated uint64 fee_tiers = 1;
+  uint64 max_JITs_per_block = 3;
```
[the remark this answers is not in the document]
Comment (NicholasDotSol) [reply docs/neutron/modules/dex/messages.md:355]: Max_JITsPerBlock is correct

File: docs/neutron/modules/dex/messages.md [hunk 7]
```diff
@@ -4,19 +4,23 @@
...
 
 ### Deposit Mechanism
 
-When depositing into an LP position (`PoolReserves`) a user specifies amounts of Token0 and Token1 as well as a TickIndex and a fee. The liquidity is added to the reserves for the respective ticks. Token0 will be deposited into the `PoolReserves` struct with the matching fee at `TickIndex - fee` and Token1 will be deposited into the `PoolReserves` at `TickIndex + fee`.
+When depositing into an LP position (`PoolReserves`) a user specifies amounts of `TokenA` and `TokenB` as well as a `TickIndexAToB` and a fee. The liquidity is added to the reserves for the respective ticks. As a convenience to users, they do not have to know the canonical `Token0` and `Token1` for a pair, instead they can assign pass either denom as `TokenA` or `TokenB`. `TokenA` and `TokenB` along with the `TickIndexAToB` will be normalized according to the canonical [`PairID`](docs/neutron/modules/dex/overview/concepts/tick-liquidity.md#pairid). Token0 will be deposited into the `PoolReserves` struct with the matching fee at `TickIndex - fee` and Token1 will be deposited into the `PoolReserves` at `TickIndex + fee`.
```
[the remark this answers is not in the document]
Comment (NicholasDotSol) [reply docs/neutron/modules/dex/messages.md:11]: TokenA and tokenB are a user convenience, the DEX internally uses token0 and token1. When we say the user doesn't need to care about the canonical toke0 and token1 we mean the user only cares about providing the correct tokenA and token B. I'll try to clarify this in a different way

File: docs/neutron/modules/dex/messages.md [hunk 8]
```diff
@@ -4,19 +4,23 @@
...
+`TickIndexAToB` specifies the tick providing the desired conversion rate from `TokenA` to `TokenB`. This means that `TokenB` will be deposited at $$TickIndexAToB + fee$$ and `TokenA` will be deposited at $$TickIndexAToB\cdot-1 + fee$$.
 
 In the most basic case, when depositing into a pool, the ratio of Token0 to Token1 will be preserved. If a user does not provide tokens in the same ratio, then only a portion of their total deposit will be used so as to maintain the pool ratio.
 
-$$true_0 = min(amountDeposited_0, \frac{existingReserves_0 \cdot amountDeposited_1}{existingReserves_1})$$ $$true_1 = min(amountDeposited_1, \frac{existingReserves_1 \cdot amountDeposited_0}{existingReserves_0})$$
+$$true_0 = min(amountDeposited_0, \frac{existingReserves_0}{existingReserves_1}\cdot amountDeposited_1)$$
+
+$$true_1 = min(amountDeposited_1, \frac{existingReserves_1}{existingReserves_0}\cdot amountDeposited_0)$$
 
 **NOTE**: Most pools will only have Token1 OR Token0, so most deposits will only be providing one tokens to one side of the pool.
```
[the remark this answers is not in the document]
Comment (NicholasDotSol) [reply docs/neutron/modules/dex/messages.md:21]: DEX deals with token0 and token1 internally

File: docs/neutron/modules/dex/messages.md [hunk 9]
```diff
@@ -26,35 +30,42 @@ $$newShares = \frac{valueDeposited \cdot totalShares}{valueTotal}$$
 
 #### Autoswap
 
-By default the `autoswap` option is enabled, which allows use to deposit their full deposit amount.  Autoswap provides a mechanism for users to deposit the entirety of their specified deposit amounts by paying a small fee. The fee for performing an autoswap is deducted from the total number of shares the the user is issued. When calculating share issuance the same formula as above is used for the balanced portion of the deposit, with the following formula use to calculate the shares issues that would unbalance the pool:
+By default the `autoswap` option is enabled, which allows users to deposit their full deposit amount.  Autoswap provides a mechanism for users to deposit the entirety of their specified deposit amounts by paying a small fee. The fee for performing an autoswap is deducted from the total number of shares the the user is issued. When calculating share issuance the same formula as above is used for the balanced portion of the deposit, with the following formula use to calculate the shares issues that would unbalance the pool:
```
[the remark this answers is not in the document]
Comment (NicholasDotSol) [reply docs/neutron/modules/dex/messages.md:33]: will add clarification

File: docs/neutron/modules/dex/overview/concepts/tick-liquidity.md [hunk 10]
```diff
@@ -1,55 +1,91 @@
...
+`TickLiquidity` structs are used to store liquidity within the Dex. Each tick has a specific price and holds liquidity for a single token. `TickLiquidity` come in two general types --`PoolReserves` for storing LP positions and `LimitOrderTranches` for storing limit orders. Both tick liquidity types are indexed by a key which shares a number of feilds in common. These two types make up the fundemental building blocks of the DEX orderbook, and are critical to the [liquidity iteration mechanism](docs/neutron/modules/dex/overview/concepts/liquidity-iteration.md). `TickLiquidity`s contain all of the neccesary information to perform a swap, namely price and available reserves. Each `TickLiquidity` instance has a `TradePairID`, `TickIndexTakerToMaker`. A `TradePairID` containing the feilds `MakerDenom` and `TakerDenom` describes what denom is held in the `TickLiquidity` (`MakerDenom`) and what denom it can be traded with (`TakerDenom`). `TickIndexTakerToMaker` describes the tick at which a certain liquidity is stored and by extension its `PriceTakerToMaker`.
```
[the remark this answers is not in the document]
Comment (NicholasDotSol) [reply docs/neutron/modules/dex/overview/concepts/tick-liquidity.md:3]: not entirely sure what you mean, attempted a fix though

File: docs/neutron/modules/dex/messages.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jcompagni10) [reply docs/neutron/modules/dex/messages.md:358]: It was changed https://github.com/neutron-org/neutron/blob/28b1d2ce968aaf1866e92d5286487f079eba3370/x/dex/types/params.pb.go#L31

**Review by NeverHappened — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Pull request merged**